### PR TITLE
Data layer refactoring

### LINKED ISSUE #355: Remove series then scroll to right after not working.
Hi, lightweight-charts team
I am facing an issue with lightweight-charts.

**Lightweight Charts Version:** 1.2.2

**Steps/code to reproduce:**

- AddLineSeries when click on button 2001 , 2002 ,2003   to add more series.
- Click a button following step on button 2003 and button 2002 to remove selected series.
- Scroll chart to the end of the right side after removing a series.

**Actual behavior:**
Show Uncaught Error: Value is null;

**Expected behavior:**
Don't show Uncaught Error: Value is null;

**Screenshots**
https://drive.google.com/file/d/1Ws0Q2y6md0ZxxpaRLrolUuBbBiZbYhup/view

**CodeSandbox/JSFiddle/etc link:**

https://stackblitz.com/edit/js-eivupz?file=index.js

### PR DESCRIPTION
**Pull request by timocov**
**Type of PR:** bugfix, enhancement

**PR checklist:**

- [x] Addresses an existing issue: fixes #355, #270
- [x] Includes tests
- [ ] Documentation update

**Overview of change:**

There is a lot of changes so I think it's better to review changes commit-by-commit.

### CODE REVIEW

File: src/gui/time-axis-widget.ts [hunk 1]
```diff
@@ -299,31 +299,12 @@ export class TimeAxisWidget implements MouseEventHandlers, IDestroyable {
...
-
-		let maxSpan = tickMarks.reduce(markWithGreaterSpan, tickMarks[0]).span;
+		let maxWeight = tickMarks.reduce(markWithGreaterWeight, tickMarks[0]).weight;
 
 		// special case: it looks strange if 15:00 is bold but 14:00 is not
-		// so if maxSpan > 30 and < 40 reduce it to 30
-		if (maxSpan > 30 && maxSpan < 40) {
-			maxSpan = 30;
+		// so if maxWeight > 30 and < 40 reduce it to 30
+		if (maxWeight > 30 && maxWeight < 40) {
```
Comment (eugene-korobko) [inline src/gui/time-axis-widget.ts:306]: If so, why do we set different weight for 1h and 3h divisors?

File: src/api/time-scale-point-weight-generator.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	{ divisor: hours(6), weight: 32 },
+	{ divisor: hours(12), weight: 33 },
+];
+
+function weightByTime(time: UTCTimestamp, prevTime: UTCTimestamp | null): number {
+	if (prevTime !== null) {
+		const prevDate = new Date(prevTime * 1000);
+		const currentDate = new Date(time * 1000);
+
+		if (currentDate.getUTCFullYear() !== prevDate.getUTCFullYear()) {
```
Comment (eugene-korobko) [inline src/api/time-scale-point-weight-generator.ts:37]: We should consider about time zones. For instance, on 1h resolution UTC midnights are highlighted, and there is no way to change it. So, the user has on way to implement "exchange time" ever with thier custom formatting function.

File: src/views/pane/histogram-pane-view.ts [hunk 3]
```diff
@@ -51,38 +49,31 @@ export class SeriesHistogramPaneView extends SeriesPaneViewBase<'Histogram', Tim
...
-			const value = barValueGetter(bar.value);
-			const paletteColorIndex = bar.value[SeriesPlotIndex.Color];
+		for (const row of this._series.bars().rows()) {
+			const value = barValueGetter(row.value);
 
-			const color = paletteColorIndex != null ? palette.colorByIndex(paletteColorIndex) : defaultColor;
-			const item = createRawItem(index, value, color);
+			const color = row.color !== undefined ? row.color : defaultColor;
+			const item = createRawItem(row.index, value, color);
 			// colorIndex is the paneview's internal palette index
```
Comment (eugene-korobko) [inline src/views/pane/histogram-pane-view.ts:67]: Is this comment still actual?

File: src/api/time-scale-point-weight-generator.ts (hunk 2 above)
Comment (timocov) [reply src/api/time-scale-point-weight-generator.ts:37]: I think it's better to discuss/try to implement in other issue. This isn't changed from master branch.

File: src/gui/time-axis-widget.ts (hunk 1 above)
Comment (timocov) [reply src/gui/time-axis-widget.ts:306]: I bet you know about that better 😂 

image

File: src/views/pane/histogram-pane-view.ts (hunk 3 above)
Comment (timocov) [reply src/views/pane/histogram-pane-view.ts:67]: Nope, will remove it.

File: src/views/pane/histogram-pane-view.ts [hunk 4]
```diff
@@ -51,38 +49,31 @@ export class SeriesHistogramPaneView extends SeriesPaneViewBase<'Histogram', Tim
...
 			// this internal palette stores defaultColor by 0 index and pallette colors by paletteColorIndex + 1
```
Comment (timocov) [inline src/views/pane/histogram-pane-view.ts:68]:
```suggestion
```

File: src/gui/time-axis-widget.ts (hunk 1 above)
Comment (timocov) [reply src/gui/time-axis-widget.ts:306]: I bet you know about that better 😂 

image

File: src/api/time-scale-point-weight-generator.ts (hunk 2 above)
Comment (timocov) [reply src/api/time-scale-point-weight-generator.ts:37]: I think it's better to discuss/try to implement in other issue. This isn't changed from master branch.

File: src/api/time-scale-point-weight-generator.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	return count * 1000;
+}
+
+interface WeightDivisor {
+	divisor: number;
+	weight: number;
+}
+
+const intradayWeightDivisors: WeightDivisor[] = [
+	{ divisor: 1, weight: 20 },
```
[the remark this answers is not in the document]
Comment (timocov) [reply src/api/time-scale-point-weight-generator.ts:21]: Interesting question. That's what we have a long time (I didn't change it in this PR):

image

¯\_(ツ)_/¯

[the remark this answers is not in the document]
Comment (eugene-korobko) [reply src/api/time-scale-point-weight-generator.ts:21]: Lets's write a TODO here...

File: src/gui/time-axis-widget.ts (hunk 1 above)
Comment (eugene-korobko) [reply src/gui/time-axis-widget.ts:306]: ok, lets keep this so far )))

File: src/api/time-scale-point-weight-generator.ts (hunk 2 above)
Comment (eugene-korobko) [reply src/api/time-scale-point-weight-generator.ts:37]: Do we have some suitible task? Or lets create new one?

Comment (timocov) [reply src/api/time-scale-point-weight-generator.ts:37]: Just created https://github.com/tradingview/lightweight-charts/issues/497

File: src/api/time-scale-point-weight-generator.ts (hunk 5 above)
Comment (timocov) [inline src/api/time-scale-point-weight-generator.ts:21]:
```suggestion
	// TODO: divisor=1 means 1ms and it's strange that weight for 1ms > weight for 1s
	{ divisor: 1, weight: 20 },
```

[the remark this answers is not in the document]
Comment (timocov) [reply src/api/time-scale-point-weight-generator.ts:22]: Done

**Pull request merged**